File: sdks/packages/typescript/examples/run-all.ts
/**
 * run-all.ts — the examples suite runner AND pre-release smoke test.
 *
 * Runs examples 01–05 in sequence against the configured daemon, sharing one
 * `IosClient`. Example 06 (UI automation) runs only when RUN_UI=1, because it
 * needs a forwarded WebDriverAgent session.
 *
 * Exit code contract (so `npm run examples` is a clean green/red check):
 *   - Every example that ran either PASSED or SKIPPED  -> exit 0 (green)
 *   - Any example threw a non-SkipError                -> exit 1 (red)
 *
 * SKIP vs FAIL:
 *   - A device-dependent example that legitimately can't run (no device
 *     attached, UI backend not forwarded) throws a `SkipError` and is reported
 *     as SKIP — it does NOT fail the suite. This keeps the smoke test honest on
 *     a daemon with no device: the reachable surface is still verified.
 *   - Any other error (auth failure, daemon down, unexpected 5xx, a bug in an
 *     example) is a FAIL and flips the exit code to 1.
 *
 * Missing GO_IOS_API_KEY is handled up front by `makeClient()` (prints help and
 * exits non-zero) before any example runs.
 *
 * Usage:
 *   GO_IOS_API_KEY=... npm run examples
 *   RUN_UI=1 GO_IOS_API_KEY=... npm run examples   # also run 06
 */
import { IosClient } from "../src/index";
import { SkipError, baseUrl, makeClient } from "./_shared";

import { run as listDevices } from "./01-list-devices";
import { run as deviceInfo } from "./02-device-info";
import { run as listApps } from "./03-list-apps";
import { run as screenshot } from "./04-screenshot";
import { run as streamSyslog } from "./05-stream-syslog";
import { run as uiAutomation } from "./06-ui-automation";

interface Example {
  readonly name: string;
  readonly run: (client: IosClient) => Promise<void>;
}

/** The ordered core suite (always runs). */
const CORE: readonly Example[] = [
  { name: "01-list-devices", run: listDevices },
  { name: "02-device-info", run: deviceInfo },
  { name: "03-list-apps", run: listApps },
  { name: "04-screenshot", run: screenshot },
  { name: "05-stream-syslog", run: streamSyslog },
];

/** Opt-in advanced example, only when RUN_UI=1. */
const UI: Example = { name: "06-ui-automation", run: uiAutomation };

type Outcome = "PASS" | "SKIP" | "FAIL";

async function runOne(client: IosClient, ex: Example): Promise<Outcome> {
  console.log(`\n=== ${ex.name} ===`);
  try {
    await ex.run(client);
    console.log(`--- ${ex.name}: PASS`);
    return "PASS";
  } catch (err) {
    if (err instanceof SkipError) {
      console.log(`--- ${ex.name}: SKIP (${err.message})`);
      return "SKIP";
    }
    console.error(
      `--- ${ex.name}: FAIL — ${err instanceof Error ? err.stack ?? err.message : String(err)}`,
    );
    return "FAIL";
  }
}

async function main(): Promise<void> {
  // `makeClient()` enforces GO_IOS_API_KEY (exits non-zero if missing).
  const client = makeClient();
  console.log(`go-ios examples suite → ${baseUrl() ?? "(auto-discovered local daemon)"}`);

  const suite: Example[] = [...CORE];
  if (process.env.RUN_UI === "1") {
    suite.push(UI);
  } else {
    console.log("(RUN_UI!=1 — skipping 06-ui-automation; set RUN_UI=1 to include it)");
  }

  const results: { name: string; outcome: Outcome }[] = [];
  for (const ex of suite) {
    const outcome = await runOne(client, ex);
    results.push({ name: ex.name, outcome });
  }

  // Summary.
  const pass = results.filter((r) => r.outcome === "PASS").length;
  const skip = results.filter((r) => r.outcome === "SKIP").length;
  const fail = results.filter((r) => r.outcome === "FAIL").length;

  console.log("\n=== summary ===");
  for (const r of results) console.log(`  ${r.outcome.padEnd(4)} ${r.name}`);
  console.log(`  ${pass} passed, ${skip} skipped, ${fail} failed`);

  // Red only if something genuinely failed; SKIPs are fine.
  process.exit(fail > 0 ? 1 : 0);
}

await main();

File: sdks/packages/python/examples/run_all.py
#!/usr/bin/env python3
"""Run every example in sequence — the pre-release smoke test.

Runs examples 01–06 against a live go-ios daemon, in order. Example 07 (UI
automation) only runs when ``RUN_UI=1`` because it needs a WebDriverAgent backend.

Exit behavior
-------------
* Any example raising an unexpected exception -> the runner prints ``FAIL`` and
  exits non-zero (so CI / a release gate catches it).
* An example that cannot run for environmental reasons (no device attached, UI
  backend unreachable) raises ``SkipExample`` -> printed as ``SKIP``; it does
  **not** fail the suite.
* Missing ``GO_IOS_API_KEY`` is reported once up front with a non-zero exit.

Run it::

    export GO_IOS_API_KEY=your-key
    # export GO_IOS_UDID=00008110-000...   # optional; first device otherwise
    uv run python examples/run_all.py
    RUN_UI=1 uv run python examples/run_all.py   # also run 07
"""

from __future__ import annotations

import importlib
import os
import sys
import traceback
from typing import Callable, List, Tuple

# Ensure this directory is importable whether run as ``python examples/run_all.py``
# or via the installed ``go-ios-examples`` console script (whose cwd is arbitrary).
_HERE = os.path.dirname(os.path.abspath(__file__))
if _HERE not in sys.path:
    sys.path.insert(0, _HERE)

from _common import SkipExample, base_url, require_api_key  # noqa: E402

# (module name, human label). Modules are imported lazily so an import error in
# one example is reported like any other failure rather than aborting the run.
_CORE: List[Tuple[str, str]] = [
    ("01_list_devices", "list devices"),
    ("02_device_info", "device info"),
    ("03_list_apps", "list apps"),
    ("04_screenshot", "screenshot"),
    ("05_stream_syslog", "stream syslog"),
    ("06_async_stream", "async stream"),
]
_UI: Tuple[str, str] = ("07_ui_automation", "ui automation")


def _load_main(module_name: str) -> Callable[[], None]:
    """Import an example module and return its ``main`` callable."""
    module = importlib.import_module(module_name)
    return module.main  # type: ignore[no-any-return]


def _run_one(module_name: str, label: str) -> str:
    """Run a single example. Returns one of ``PASS`` / ``SKIP`` / ``FAIL``."""
    print(f"\n########## {module_name} — {label} ##########")
    try:
        _load_main(module_name)()
    except SkipExample as skip:
        print(f"SKIP: {module_name}: {skip}")
        return "SKIP"
    except Exception:  # noqa: BLE001 — the runner is a top-level harness
        print(f"FAIL: {module_name}:")
        traceback.print_exc()
        return "FAIL"
    print(f"PASS: {module_name}")
    return "PASS"


def main() -> int:
    # Fail fast and clearly if the daemon key is missing (one message, not six).
    require_api_key()
    print(f"go-ios-sdk examples smoke test against {base_url() or '(auto-discovered daemon)'}")

    plan = list(_CORE)
    if os.environ.get("RUN_UI") == "1":
        plan.append(_UI)
    else:
        print("(RUN_UI!=1; skipping 07_ui_automation — set RUN_UI=1 to include it)")

    results = [(name, _run_one(name, label)) for name, label in plan]

    print("\n========== summary ==========")
    for name, status in results:
        print(f"  {status:4s}  {name}")

    failed = [name for name, status in results if status == "FAIL"]
    if failed:
        print(f"\n{len(failed)} example(s) FAILED: {', '.join(failed)}")
        return 1
    print("\nall examples passed (or skipped gracefully)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
